ext/standard/array.c

```diff
@@ -3507,51 +3532,55 @@ PHP_FUNCTION(array_slice)
 	/* Initialize returned array */
 	array_init_size(return_value, (uint32_t)length);
 
-	/* Start at the beginning and go until we hit offset */
-	pos = 0;
-	if (HT_IS_PACKED(Z_ARRVAL_P(input)) &&
-	    (!preserve_keys ||
-	     (offset == 0 && HT_IS_WITHOUT_HOLES(Z_ARRVAL_P(input))))) {
-		zend_hash_real_init_packed(Z_ARRVAL_P(return_value));
-		ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(return_value)) {
-			ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(input), entry) {
-				pos++;
-				if (pos <= offset) {
-					continue;
-				}
-				if (pos > offset + length) {
-					break;
-				}
-				if (UNEXPECTED(Z_ISREF_P(entry)) &&
-					UNEXPECTED(Z_REFCOUNT_P(entry) == 1)) {
-					entry = Z_REFVAL_P(entry);
-				}
-				Z_TRY_ADDREF_P(entry);
-				ZEND_HASH_FILL_ADD(entry);
-			} ZEND_HASH_FOREACH_END();
-		} ZEND_HASH_FILL_END();
-	} else {
-		ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(input), num_key, string_key, entry) {
-			pos++;
-			if (pos <= offset) {
-				continue;
-			}
-			if (pos > offset + length) {
-				break;
-			}
-
-			if (string_key) {
-				entry = zend_hash_add_new(Z_ARRVAL_P(return_value), string_key, entry);
-			} else {
-				if (preserve_keys) {
-					entry = zend_hash_index_add_new(Z_ARRVAL_P(return_value), num_key, entry);
-				} else {
-					entry = zend_hash_next_index_insert_new(Z_ARRVAL_P(return_value), entry);
-				}
-			}
-			zval_add_ref(entry);
-		} ZEND_HASH_FOREACH_END();
-	}
+    // Contains modified variants of ZEND_HASH_FOREACH_VAL
+    {
+        HashTable *ht = Z_ARRVAL_P(input);
+        Bucket *p = find_bucket_at_offset(ht, offset);
+        Bucket *end = ht->arData + ht->nNumUsed;
+
+        /* Start at the beginning and go until we hit offset */
+        if (HT_IS_PACKED(Z_ARRVAL_P(input)) &&
+            (!preserve_keys ||
+             (offset == 0 && HT_IS_WITHOUT_HOLES(Z_ARRVAL_P(input))))) {
+
+            zend_hash_real_init_packed(Z_ARRVAL_P(return_value));
+            ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(return_value)) {
+                for (; p != end; p++) {
```

Comment:
Both of these loops no longer skip over undef elements.

Reply:
Fixed and added a new test